### PR TITLE
add operational model for pthread_detach

### PR DESCRIPTION
**Pull request by fatimahkj**
This PR contains the operational model for the pthread_detach function. 

Signed-off-by: fatimahkj <fatimahaljaafari@gmail.com>

### CODE REVIEW

File: src/c2goto/library/pthread_lib.c [hunk 1]
```diff
@@ -275,6 +278,23 @@ __ESBMC_HIDE:;
...
+// source: https://man7.org/linux/man-pages/man3/pthread_detach.3.html
+int pthread_detach(pthread_t threadid)
+{
+__ESBMC_HIDE:;
+  __ESBMC_atomic_begin();
+
+  _Bool det = __ESBMC_pthread_thread_detach[(int)threadid];
+  // abort execution when thread is detached;
+  __ESBMC_assert(
+    det != 0, "You want to detach a thread which already detached");
```
Comment (rafaelsamenezes) [inline src/c2goto/library/pthread_lib.c]: Just a nitpick here. You are declaring the variable `det` that is only once used, use the value directly. I would change the error message to:

"Detaching a detached thread is Undefined Behaviour"

File: src/c2goto/library/pthread_lib.c [hunk 2]
```diff
@@ -275,6 +278,23 @@ __ESBMC_HIDE:;
...
+  __ESBMC_terminate_thread();
+  __ESBMC_pthread_thread_detach[(int)threadid] = 1;
+  __ESBMC_atomic_end();
+
+  return 0;
```
Comment (rafaelsamenezes) [inline src/c2goto/library/pthread_lib.c]: The reference defines three return values:

- 0: When no error occurred
- EINVAL:  thread is not a joinable thread.
- ESRCH:  No thread with the ID thread could be found.

**Review by rafaelsamenezes — CHANGES_REQUESTED** [review]

**Comment by lucasccordeiro** [comment]
@fatimahkj: Test case `01_pthread67` is failing in our CI. Could you please check it?

**Review by lucasccordeiro — CHANGES_REQUESTED** [review]
I think we should have something like:

```
// The pthread_detach() marks the thread identified by thread as detached.
// source: https://man7.org/linux/man-pages/man3/pthread_detach.3.html
int pthread_detach(pthread_t threadid)
{
__ESBMC_HIDE:;
  __ESBMC_atomic_begin();

  if(__ESBMC_pthread_thread_detach[(int)threadid])
    return EINVAL; //  thread is not a joinable thread.
  if(!__ESBMC_pthread_thread_running[(int)threadid])
    return ESRCH; // No thread with the ID thread could be found.
  __ESBMC_terminate_thread();
  __ESBMC_pthread_thread_detach[(int)threadid] = 1;
  __ESBMC_atomic_end();

  return 0; // no error occurred
}
```

**Comment by lucasccordeiro** [comment]
@fatimahkj: I believe you would need to write five test cases to check your implementation as follows:

thread is not a joinable thread.
```
assert(pthread_detach(some_thread_id) == EINVAL)
```

thread is a joinable thread.
```
assert(pthread_detach(some_thread_id) != EINVAL)
```

No thread with the ID thread could be found.
```
assert(pthread_detach(some_thread_id) == ESRCH)
```

Thread with the ID thread could be found.
```
assert(pthread_detach(some_thread_id) != ESRCH)
```

// no error occurred
```
assert(pthread_detach(some_thread_id) == 0)
```

**Comment by lucasccordeiro** [comment]
Closed via https://github.com/esbmc/esbmc/pull/243.

**Pull request closed without merging**